### PR TITLE
CRM-20199 Contribution search performance

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
* [CRM-19830: Cleanup contribution search by making non-exportable fields exportable](https://issues.civicrm.org/jira/browse/CRM-19830)

---

 * [CRM-19815: Remove performance degrading-joins from civicrm_contribution search](https://issues.civicrm.org/jira/browse/CRM-19815)

---

 * [CRM-20199: Contribution searches all include an unindexed join](https://issues.civicrm.org/jira/browse/CRM-20199)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@colemanw @monishdeb FYI - I am trying to clean up the contribution search class a bit in order to fix some of the performance issues. I dug into why some fields were manually declared in the hope of being a bit more meta-data-ey & found that it was because they were set to export = false. In each case a whole lot of handling had been added to compensate for it so I think we can add export = true more aggressively.

I'd like to remove the handling for contribution_check_number but it seems that as a result of it being there contribution_check_number is accepted & output (which is used by search fields & could possibly conflict with trxn_id I guess.....). I think the hated uniquenames are the right answer here + add a bit of handling in the contribution.get api to copy $result['contribution_check_number' to $result['check_number'] - this doesn't happen currently for contribution_source but perhaps it should. 

That + dealing properly with the pseudoconstant fields (payment_instrument, financial_type, contribution_status is next)

Contribution is of course one of a handful of api using the query object....

File: CRM/Contribute/BAO/Query.php [hunk 1]
```diff
@@ -157,100 +151,12 @@ public static function select(&$query) {
       $query->_tables['civicrm_contribution'] = 1;
     }
 
-    // LCD 716
-    $includeSoftCredits = self::isSoftCreditOptionEnabled($query->_params);
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: This is just moved

File: CRM/Contribute/BAO/Query.php [hunk 2]
```diff
@@ -157,100 +151,12 @@ public static function select(&$query) {
...
-      $query->_tables['civicrm_contribution_soft'] = 1;
-    }
-
     if (!empty($query->_returnProperties['contribution_campaign_title'])) {
       $query->_select['contribution_campaign_title'] = "civicrm_campaign.title as contribution_campaign_title";
       $query->_element['contribution_campaign_title'] = $query->_tables['civicrm_campaign'] = 1;
     }
 
-    // Adding address_id in a way that is more easily extendable since the above is a bit ... wordy.
-    $supportedBasicReturnValues = array('address_id');
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: gone

File: CRM/Contribute/BAO/Query.php [hunk 3]
```diff
@@ -157,100 +151,12 @@ public static function select(&$query) {
...
-    foreach ($supportedBasicReturnValues as $fieldName) {
-      if (!empty($query->_returnProperties[$fieldName]) && empty($query->_select[$fieldName])) {
-        $query->_select[$fieldName] = "civicrm_contribution.{$fieldName} as $fieldName";
-        $query->_element[$fieldName] = $query->_tables['civicrm_contribution'] = 1;
-      }
-    }
-
-    //CRM-16116: get financial_type_id
-    if (!empty($query->_returnProperties['financial_type_id'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: not needed now the field is set to export

File: CRM/Contact/BAO/Query.php [hunk 4]
```diff
@@ -624,22 +632,28 @@ public function addSpecialFields($apiEntity) {
...
     foreach ($this->_fields as $name => $field) {
       // skip component fields
       // there are done by the alter query below
       // and need not be done on every field
+      // @todo remove these & handle using metadata - only obscure fields
+      // that are hack-added should need to be excluded from the main loop.
       if (
         (substr($name, 0, 12) == 'participant_') ||
         (substr($name, 0, 7) == 'pledge_') ||
-        (substr($name, 0, 5) == 'case_') ||
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: ie do not skip the loop for fields starting with contribution_ as many of them are in fact better dealt with in that context since they can be derived from metadata

File: CRM/Contact/BAO/Query.php [hunk 5]
```diff
@@ -680,15 +697,12 @@ public function selectClause($apiEntity = NULL) {
         }
       }
 
-      if (in_array($name, array('prefix_id', 'suffix_id', 'gender_id', 'communication_style_id'))) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: Consolidate this logic with general option-value pseudoconstant metadata based evaluation further down

File: CRM/Contact/BAO/Query.php [hunk 6]
```diff
@@ -680,15 +697,12 @@ public function selectClause($apiEntity = NULL) {
...
-        }
-      }
-
       $cfID = CRM_Core_BAO_CustomField::getKeyID($name);
-      if (!empty($this->_paramLookup[$name]) || !empty($this->_returnProperties[$name]) ||
-        $makeException
+      if (
+        !empty($this->_paramLookup[$name])
+        || !empty($this->_returnProperties[$name])
+        || $this->pseudoConstantNameIsInReturnProperties($field)
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: $this->pseudoConstantNameIsInReturnProperties replaces the earlier in_array call

File: CRM/Contact/BAO/Query.php [hunk 7]
```diff
@@ -805,42 +822,9 @@ public function selectClause($apiEntity = NULL) {
                 $this->_element['provider_id'] = 1;
               }
 
-              if ($tName == 'contact') {
+              if ($tName == 'contact' && $fieldName == 'organization_name') {
                 // special case, when current employer is set for Individual contact
-                if ($fieldName == 'organization_name') {
-                  $this->_select[$name] = "IF ( contact_a.contact_type = 'Individual', NULL, contact_a.organization_name ) as organization_name";
-                }
-                elseif ($fieldName != 'id') {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: move all this stuff out of the 'contact-table-specific' & make it just 'metadata-based'

File: CRM/Contact/BAO/Query.php [hunk 8]
```diff
@@ -856,7 +840,20 @@ public function selectClause($apiEntity = NULL) {
                 }
               }
               else {
-                $this->_select[$name] = "{$field['where']} as `$name`";
+                // If we have an option group defined then rather than joining the option value table in
+                // (which is an unindexed join) we render the option value on output.
+                // @todo - extend this to other pseudoconstants.
+                if ($this->pseudoConstantNameIsInReturnProperties($field)) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: By adding all option-value based fields (that get here) to the pseudoconstant array they become workable in the context of the $convertToPseudoFields fn

File: CRM/Contact/BAO/Query.php [hunk 9]
```diff
@@ -6136,62 +6161,91 @@ protected function prepareOrderBy($sort, $sortByChar, $sortOrder, $additionalFro
...
       }
       elseif ($sortByChar) {
-        $order = " ORDER BY UPPER(LEFT(contact_a.sort_name, 1)) asc";
+        $orderByArray = array("UPPER(LEFT(contact_a.sort_name, 1)) asc");
       }
       else {
         $order = " ORDER BY contact_a.sort_name asc, contact_a.id";
       }
     }
+    if (!$order && empty($orderByArray)) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: Refactoring in this function ensures an array of order-by fields although it is still pretty funky. Previously the function did not iterate through all fields - just the first one in the string, to add special handling.

The handling we are adding is a way of avoiding joining the option_value table in when sorting by it. By rendering the values in php & using field($fieldName, 4,6,7) I was able to render the query without the join. Results in test were 10 times faster, in some cases much more.

File: CRM/Contribute/BAO/Contribution.php [hunk 10]
```diff
@@ -825,15 +838,6 @@ public static function &exportableFields($checkPermission = TRUE) {
         ),
       );
 
-      $contributionRecurId = array(
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: As part of this PR I declared this field as exportable so it no longer needs manual adding to exportable fields

File: CRM/Contribute/BAO/Contribution.php [hunk 11]
```diff
@@ -781,32 +781,45 @@ public static function &importableFields($contactType = 'Individual', $status =
...
       $impFields = CRM_Contribute_DAO_Contribution::export();
+      foreach ($impFields as &$field) {
+        // @todo consider declaring this at the DAO level.
+        $field['bao'] = 'CRM_Contribute_BAO_Contribution';
+      }
       $expFieldProduct = CRM_Contribute_DAO_Product::export();
       $expFieldsContrib = CRM_Contribute_DAO_ContributionProduct::export();
       $typeField = CRM_Financial_DAO_FinancialType::export();
       $financialAccount = CRM_Financial_DAO_FinancialAccount::export();
-      $optionField = CRM_Core_OptionValue::getFields($mode = 'contribute');
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: as part of this PR I declared payment_instrument_id as exportable so it no longer needs manual adding

File: CRM/Contribute/BAO/Contribution.php [hunk 12]
```diff
@@ -781,32 +781,45 @@ public static function &importableFields($contactType = 'Individual', $status =
...
-      $contributionStatus = array(
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: now handled by the pseudoconstant - not needed as a manual add on

File: CRM/Contribute/BAO/Contribution.php [hunk 13]
```diff
@@ -3711,6 +3717,9 @@ public static function deleteContactContribution($contactId) {
   public static function buildOptions($fieldName, $context = NULL, $props = array()) {
     $className = __CLASS__;
     $params = array();
+    if (isset($props['orderColumn'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: Allow a request for orderBy to be passed through buildOptions

File: CRM/Contribute/BAO/Query.php [hunk 14]
```diff
@@ -32,36 +32,27 @@
...
    * @return array
    *   Associative array of contribution fields
    */
   public static function getFields($checkPermission = TRUE) {
-    if (!self::$_contributionFields) {
-      self::$_contributionFields = array();
-
-      $fields = CRM_Contribute_BAO_Contribution::exportableFields($checkPermission);
-
+    if (!isset(\Civi::$statics[__CLASS__]) || !isset(\Civi::$statics[__CLASS__]['fields']) || !isset(\Civi::$statics[__CLASS__]['contribution'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: use \Civi::$statics for statics as a preferred style thing

File: CRM/Contribute/BAO/Query.php [hunk 15]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
       $query->_tables['contribution_batch'] = 1;
     }
 
-    if (!empty($query->_returnProperties['contribution_source'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: now declared as exportable so handled by metadata

File: CRM/Contribute/BAO/Query.php [hunk 16]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
...
-      $query->_select['contribution_source'] = "civicrm_contribution.source as contribution_source";
-      $query->_element['contribution_source'] = 1;
-      $query->_tables['civicrm_contribution'] = 1;
-    }
-
-    // get contribution_status
-    if (!empty($query->_returnProperties['contribution_status_id'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: now handled by metadata as main loop does not skip contribution_ fields

File: CRM/Contribute/BAO/Query.php [hunk 17]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
...
-      $query->_select['contribution_status_id'] = "contribution_status.value as contribution_status_id";
-      $query->_element['contribution_status_id'] = 1;
-      $query->_tables['civicrm_contribution'] = 1;
-      $query->_tables['contribution_status'] = 1;
-    }
-
-    // get contribution_status label
-    if (!empty($query->_returnProperties['contribution_status'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: now handled by metadata as  a pseudoconstant

File: CRM/Contribute/BAO/Query.php [hunk 18]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
...
-      $query->_select['contribution_status'] = "contribution_status.label as contribution_status";
-      $query->_element['contribution_status'] = 1;
-      $query->_tables['civicrm_contribution'] = 1;
-      $query->_tables['contribution_status'] = 1;
-    }
-
-    // get payment instrument
-    if (!empty($query->_returnProperties['payment_instrument'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: now handled by metadata as main loop does not skip payment_ fields & has been changed to exportable & handled via metadata

File: CRM/Contribute/BAO/Query.php [hunk 19]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
...
-    // get payment instrument id
-    if (!empty($query->_returnProperties['payment_instrument_id'])) {
-      $query->_select['instrument_id'] = "contribution_payment_instrument.value as instrument_id";
-      $query->_select['payment_instrument_id'] = "contribution_payment_instrument.value as payment_instrument_id";
-      $query->_element['instrument_id'] = $query->_element['payment_instrument_id'] = 1;
-      $query->_tables['civicrm_contribution'] = 1;
-      $query->_tables['contribution_payment_instrument'] = 1;
-    }
-
-    if (!empty($query->_returnProperties['check_number'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: now declared as exportable and handled as metadata. Api has a specific hack to provide consistent return for check_number vs contribution_check_number

File: CRM/Contribute/BAO/Query.php [hunk 20]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
...
-      $query->_select['contribution_check_number'] = "civicrm_contribution.check_number as contribution_check_number";
-      $query->_element['contribution_check_number'] = 1;
-      $query->_tables['civicrm_contribution'] = 1;
-    }
-
-    if (!empty($query->_returnProperties['contribution_campaign_id'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: now handled by metadata as main loop does not skip contribution_ fields

File: CRM/Contribute/BAO/Query.php [hunk 21]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
...
-      $query->_select['contribution_campaign_id'] = 'civicrm_contribution.campaign_id as contribution_campaign_id';
-      $query->_element['contribution_campaign_id'] = 1;
-      $query->_tables['civicrm_contribution'] = 1;
-    }
-
-    // LCD 716
-    $includeSoftCredits = self::isSoftCreditOptionEnabled($query->_params);
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: just moved to a separate function for readability (ignorability). Let's face it no-one ever wrote tests for soft credit stuff....

File: CRM/Contribute/BAO/Query.php [hunk 22]
```diff
@@ -106,151 +97,13 @@ public static function select(&$query) {
...
-    foreach ($supportedBasicReturnValues as $fieldName) {
-      if (!empty($query->_returnProperties[$fieldName]) && empty($query->_select[$fieldName])) {
-        $query->_select[$fieldName] = "civicrm_contribution.{$fieldName} as $fieldName";
-        $query->_element[$fieldName] = $query->_tables['civicrm_contribution'] = 1;
-      }
-    }
+    self::addSoftCreditFields($query);
 
-    //CRM-16116: get financial_type_id
-    if (!empty($query->_returnProperties['financial_type_id'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: now handled by metadata

File: CRM/Contribute/BAO/Query.php [hunk 23]
```diff
@@ -282,7 +135,6 @@ public static function where(&$query) {
           continue;
         }
 
-        $grouping = $query->_params[$id][3];
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: unused

File: CRM/Contribute/BAO/Query.php [hunk 24]
```diff
@@ -298,7 +150,7 @@ public static function whereClauseSingle(&$values, &$query) {
     list($name, $op, $value, $grouping, $wildcard) = $values;
 
     $quoteValue = NULL;
-    $fields = array_merge(CRM_Contribute_BAO_Contribution::fields(), self::getFields());
+    $fields = self::getFields();
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: this was just to add in the missing payment_instrument_id - I added that to the exportables & so not needed

File: CRM/Contribute/BAO/Query.php [hunk 25]
```diff
@@ -313,6 +165,19 @@ public static function whereClauseSingle(&$values, &$query) {
         return;
       }
     }
+    // These are legacy names.
+    // @todo enotices when these are hit so we can start to elimnate them.
+    $fieldAliases = array(
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: just doing this to simplify the switch statement. Semi-unrelated refactor

File: CRM/Contribute/BAO/Query.php [hunk 26]
```diff
@@ -386,21 +246,14 @@ public static function whereClauseSingle(&$values, &$query) {
...
-              'contribution_source',
-              'contribution_trxn_id',
-              'contribution_check_number',
-              'contribution_payment_instrument_id',
-              'contribution_contact_id',
-              'contribution_campaign_id',
-            )
-          )
+
+        $fieldNamesNotToStripContributionFrom = array(
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: simplify the above if by opt in rather than out

Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: Note we could use 'presence in $this->fields' as a metadata based way to opt in or out

File: CRM/Contribute/BAO/Query.php [hunk 27]
```diff
@@ -693,18 +546,6 @@ public static function from($name, $mode, $side) {
         $from .= " $side  JOIN civicrm_product ON civicrm_contribution_product.product_id =civicrm_product.id ";
         break;
 
-      case 'contribution_payment_instrument':
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: Yay - go bad join go

File: CRM/Contribute/BAO/Query.php [hunk 28]
```diff
@@ -937,11 +778,11 @@ public static function defaultReturnProperties($mode, $includeCustomFields = TRU
         // torture
         'accounting_code' => 1,
         // small
-        'payment_instrument' => 1,
+        // 'payment_instrument' => 1,
         // kittens
         'payment_instrument_id' => 1,
         // argh
-        'check_number' => 1,
+        'contribution_check_number' => 1,
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: yep - that's what unique fields does for you. Control your hatred @colemanw  - it does simplify things really. Although I had to add some legacy handling to the api

File: CRM/Contribute/BAO/Query.php [hunk 29]
```diff
@@ -1207,4 +1048,86 @@ public static function formRule($fields, $files, $form) {
...
   }
 
+  /**
+   * Add the soft credit fields to the select fields.
+   *
+   * Extracted into separate function to improve readability of main select function.
+   *
+   * @param $query
+   */
+  private static function addSoftCreditFields(&$query) {
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: just moved for readability

File: CRM/Contribute/DAO/Contribution.php [hunk 30]
```diff
@@ -283,9 +283,9 @@ static function &fields() {
         'financial_type_id' => array(
           'name' => 'financial_type_id',
           'type' => CRM_Utils_Type::T_INT,
-          'title' => ts('Financial Type') ,
+          'title' => ts('Financial Type ID') ,
```
Comment (eileenmcnaughton) [inline CRM/Contribute/DAO/Contribution.php]: seems more consistent with other fields

File: CRM/Contribute/DAO/Contribution.php [hunk 31]
```diff
@@ -322,8 +322,12 @@ static function &fields() {
         'payment_instrument_id' => array(
           'name' => 'payment_instrument_id',
           'type' => CRM_Utils_Type::T_INT,
-          'title' => ts('Payment Method') ,
+          'title' => ts('Payment Method ID') ,
           'description' => 'FK to Payment Instrument',
+          'export' => true,
```
Comment (eileenmcnaughton) [inline CRM/Contribute/DAO/Contribution.php]: honestly - the way the code works we need to really really NOT want to expose stuff before we don't say export = true. api_key is a good example of when we should use false...

File: CRM/Contribute/DAO/Contribution.php [hunk 32]
```diff
@@ -609,14 +617,18 @@ static function &fields() {
             'optionEditPath' => 'civicrm/admin/options/contribution_status',
           )
         ) ,
-        'address_id' => array(
+        'contribution_address_id' => array(
```
Comment (eileenmcnaughton) [inline CRM/Contribute/DAO/Contribution.php]: yep - add unique name or it clobbers contact address_id (it kind of already was doing that ....)

File: CRM/Contribute/Selector/Search.php [hunk 33]
```diff
@@ -366,6 +366,7 @@ public function &getRows($action, $offset, $rowCount, $sort, $output = NULL) {
     $allCampaigns = CRM_Campaign_BAO_Campaign::getCampaigns(NULL, NULL, FALSE, FALSE, FALSE, TRUE);
 
     while ($result->fetch()) {
+      $this->_query->convertToPseudoNames($result);
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Selector/Search.php]: Yay. Adds the contribution_status = 'Completed' to the array if contribution_status_id = 1 exists

File: CRM/Core/BAO/Mapping.php [hunk 34]
```diff
@@ -1126,7 +1126,7 @@ public static function &returnProperties(&$params) {
    * @return NULL
    */
   public static function saveMappingFields(&$params, $mappingId) {
-    //delete mapping fields records for exixting mapping
+    //delete mapping fields records for existing mapping
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/Mapping.php]: just 'cos

File: CRM/Core/BAO/Mapping.php [hunk 35]
```diff
@@ -381,7 +381,7 @@ public static function buildMappingForm(&$form, $mappingType, $mappingId, $colum
 
     if (($mappingType == 'Search Builder') || ($exportMode == CRM_Export_Form_Select::CONTRIBUTE_EXPORT)) {
       if (CRM_Core_Permission::access('CiviContribute')) {
-        $fields['Contribution'] = CRM_Contribute_BAO_Contribution::exportableFields();
+        $fields['Contribution'] = CRM_Contribute_BAO_Contribution::getExportableFieldsWithPseudoConstants();
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/Mapping.php]: This means we see both Contribution Status & Contribution Status ID in search builder. It is consistent with pre-existing behaviour

File: CRM/Core/BAO/OptionGroup.php [hunk 36]
```diff
@@ -184,4 +184,39 @@ public static function ensureOptionGroupExists($params) {
     }
   }
 
+  /**
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/OptionGroup.php]: opportunity to add loads of lines & get a really low add vs delete score

File: CRM/Utils/System.php [hunk 37]
```diff
@@ -1416,7 +1416,7 @@ public static function flushCache() {
     CRM_Contact_BAO_Contact::$_importableFields = CRM_Contact_BAO_Contact::$_exportableFields
       = CRM_Contribute_BAO_Contribution::$_importableFields
         = CRM_Contribute_BAO_Contribution::$_exportableFields
-          = CRM_Pledge_BAO_Pledge::$_exportableFields = CRM_Contribute_BAO_Query::$_contributionFields
```
Comment (eileenmcnaughton) [inline CRM/Utils/System.php]: now in Civi::statics so otherwise purged. Not on it's own line so no point for me

File: CRM/Utils/Type.php [hunk 38]
```diff
@@ -290,7 +290,33 @@ public static function escape($data, $type, $abort = TRUE) {
       case 'MysqlOrderBy':
         if (CRM_Utils_Rule::mysqlOrderBy($data)) {
           $parts = explode(',', $data);
-          foreach ($parts as &$part) {
+
+          // The field() syntax is tricky here because it uses commas & when
```
Comment (eileenmcnaughton) [inline CRM/Utils/Type.php]: comments lose you points

File: api/v3/utils.php [hunk 39]
```diff
@@ -520,9 +520,14 @@ function _civicrm_api3_get_using_query_object($entity, $params, $additional_opti
     CRM_Utils_Array::value('return', $options, array()),
     CRM_Utils_Array::value('return', $additional_options, array())
   );
+  if (isset($returnProperties['id']) && $entity != 'Contact') {
```
Comment (eileenmcnaughton) [inline api/v3/utils.php]: because return id means return contact_id so change to contribution_id

**Comment by eileenmcnaughton** [comment]
OK this is passing & ready for QA. I've added a lot of line notes....

This basically eliminates the join to the option value table when filtering, selecting, or ordering by contribution_status_id & payment_instrument_id. In my tests this gave significant performance increases. 

In doing this PR I moved some logic from hard-coded to metadata based & where possible I moved hacky handling from the query layer further out to the api layer. Unfortunately the uniquefields stuff really is what works at the query layer so I wrote a test to make sure the same return values were accepted & returned before & after the changes & did such hackery as is required at the api layer. I think we are getting close to the point where the contribution api could use basic get & then hack in on top legacy stuff. Breaking the connection between the api & the query object feels like a good thing.

**Comment by eileenmcnaughton** [comment]
In terms of verifying the impact of this I have been checking search builder filters, returns, displays and sorts on the fields gender_id, prefix_id, contribution_status_id, contribution_status (shows labels), payment_instrument_id, payment_instrument, check_number, communication_style & communication_method as these are all fields that are touched by the changes in some way

I have also added an api test that checks that returnable fields are still returnable after the change.

The 2 screenshots here are search builder - effectively 'before' & after'

Before
screenshot 2017-01-11 04 09 08

After
screenshot 2017-01-11 04 08 53

**Comment by eileenmcnaughton** [comment]
OK just a quick risk analysis. contribution.check_number now has a uniquename as does contribution.address_id. I think the former would work uninterupted in a smart group but I should double check.

preferred_communication_method is now rendered into an array in the query object pseudoconstant handler & I have had to remove the other converters from that field. Otherwise I think the changes made to the other fields (payment_instrument_id, contribution_status_id) should only add to the output & gender, prefix_id, suffix_id should be unchanged after some mickey-mouse discovery on them.

I think testing a couple of smart groups & some profiles is probably the main need now.

**Comment by seamuslee001** [comment]
@eileenmcnaughton I have just done a test on export field mapping using check number field and it did not export

**Comment by seamuslee001** [comment]
I tested loading a saved smart group search criteria and worked fine and also all other fields in the pre saved export mapping worked

**Comment by eileenmcnaughton** [comment]
@seamuslee001 so, I feel like the answer is to run an upgrade script "UPDATE civicrm_uf_field SET field_name = 'contribution_check_number' WHERE field_name = 'check_number'

Do you think that will sort it?

**Comment by seamuslee001** [comment]
Maybe but wouldn't this be in the mapping_field table? as its a saved export field mapping?

**Comment by eileenmcnaughton** [comment]
Ah, yes

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I just added an upgrade line

**Comment by eileenmcnaughton** [comment]
I tested doing a financial batch data entry for a contribution, including payment_method & check number & it worked

**Comment by eileenmcnaughton** [comment]
I just discussed with @mlutfy & he didn't see any cause for l10n on this commit.

**Comment by seamuslee001** [comment]
@eileenmcnaughton confirmed your latest update fixes the check issue for me

File: CRM/Contribute/BAO/Contribution.php [hunk 40]
```diff
@@ -781,32 +781,45 @@ public static function &importableFields($contactType = 'Individual', $status =
...
    *
    * @param bool $checkPermission
    *
    * @return array
    *   array of exportable Fields
    */
-  public static function &exportableFields($checkPermission = TRUE) {
+  public static function exportableFields($checkPermission = TRUE) {
     if (!self::$_exportableFields) {
       if (!self::$_exportableFields) {
```
Comment (ErichBSchulz) [inline CRM/Contribute/BAO/Contribution.php]: eh?

**Review by ErichBSchulz — COMMENTED** [review]
I'm not accross this code at all - but spotted one `if` that looked redundant

**Comment by eileenmcnaughton** [comment]
Do you feel it's good to merge? Any concerns @monishdeb @colemanw

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this is the remaining work to remove the slow joins. I don't feel I can break it into any smaller chunks & I do think this is a good time of the cycle to merge it.

From the commit notes
    This involves setting payment instrument to exportable & remove hacks that were used in
    place of doing that, and adding new logic to the order by to allow it to order without
    joining the option value table (using the field() function).
    
    I have observed one small regression which is the Qill showing 'Payment Method ID'
    instead of 'Payment Method'. I'm committed to fixing that but I would push
    to merge this in advance of that because we are in the best place in the monthly
    cycle to merge this and that is a very minor fallout.

**Comment by seamuslee001** [comment]
@eileenmcnaughton i have tested this on export and search and all appears fine. I think we can address the Qill issue in another PR

**Comment by seamuslee001** [comment]
I did an export of a saved field mapping, There were no signs of fields missing and then on export fields were correctly populated as expected

**Comment by eileenmcnaughton** [comment]
:-)
